google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -3343,6 +3451,85 @@ private void createSessions(final int sessionCount, boolean distributeOverChanne
     }
   }
 
+  /**
+   * Callback interface which is invoked when a multiplexed session is being replaced by the
+   * background maintenance thread. When a multiplexed session creation fails during background
+   * thread, it would simply log the exception and retry the session creation in the next background
+   * thread invocation.
+   *
+   * <p>This consumer is not used when the multiplexed session is getting initialized for the first
+   * time during application startup. We instead use {@link
+   * MultiplexedSessionInitializationConsumer} for the first time when multiplexed session is
+   * getting created.
+   */
+  class MultiplexedSessionMaintainerConsumer extends MultiplexedSessionInitializationConsumer {
+
+    /**
+     * Method which logs the exception so that session creation can be re-attempted in the next
+     * background thread invocation.
+     */
+    @Override
+    public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount) {
+      synchronized (lock) {
+        multiplexedSessionBeingCreated = false;
+      }
+      logger.log(
+          Level.WARNING,
+          String.format(
+              "Failed to create multiplexed session. "
+                  + "Pending replacing stale multiplexed session",
+              t));
+    }
+  }
+
+  /**
+   * Callback interface which is invoked when a multiplexed session is getting initialised for the
+   * first time when a session is getting created.
+   */
+  class MultiplexedSessionInitializationConsumer implements SessionConsumer {
+    @Override
+    public void onSessionReady(SessionImpl sessionImpl) {
+      final SettableFuture settableFuture = SettableFuture.create();
```

Comment:
1. Could we split the logic of 'create the first multiplexed session' and 'replace the existing multiplexed session' into the two sub-classes that we actually have? This is confusing, as the Javadoc says that this is only used for the initial creation.
2. I think that the current implementation creates a problem for threads that wait for the initial creation of a multiplexed session. Assume that the following happens:
 2.1. A `DatabaseClient` is created (currently non-blocking). The initial value of `currentMultiplexedSessionReference` is set to a `SettableApiFuture.create()` in the constructor.
2.2. A thread tries to get a multiplexed session before it is done. It is then kept waiting on line 3001 (`currentMultiplexedSessionReference.get().get()`). That line will use the `SettableApiFuture` that was created in the constructor.
2.3. This implementation just swaps the `SettableApiFuture` in the `currentMultiplexedSessionReference` with a new instance, which means that the initial `SettableApiFuture` will never finish. That again means that those initial threads will be blocked forever.

Reply:
The fact that both this problem, but also the one mentioned further above regarding the `singleUseIsNonBlocking()` test, seem to introduce problems that are not caught by the current tests, indicate that we should enable multiplexed sessions by default for this PR, so we can actually test the implementation. We can always flip the flag back to false once we feel confident that it works before actually merging it.

Reply:
* Done. Split the logic to two different implementations. For initialisation, I am doing a get() and then set() on the `SettableApiFuture` object. That should fix the use-case you mentioned.
* For maintainer threads, get() / set() wont work since the Future is already set. Hence, swapping with a new Future.
* Regarding tests, I will use this single PR - https://github.com/googleapis/java-spanner/pull/3016 for the remainder of refactoring and parameterizing the tests.